controllers/storagecluster/external_ocs.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		s, ok := status.FromError(err)
+		if !ok {
+			// This is not gRPC error. The operation must have failed before gRPC
+			// method was called, otherwise we would get gRPC error.
+			return nil, reconcile.Result{}, err
+		}
+		r.logGrpcErrorAndReportEvent(instance, "GetStorageConfig", err, s.Code())
+
+		// storage consumer is not ready yet, requeue after some time
+		if s.Code() == codes.Unavailable {
+			return nil, reconcile.Result{RequeueAfter: time.Second * 5}, nil
+		}
+
+		return nil, reconcile.Result{}, err
+	}
+
+	var externalResources []ExternalResource
+	err = json.Unmarshal(response.Data, &externalResources)
+	if err != nil {
+		return nil, reconcile.Result{}, err
+	}
+
+	return externalResources, reconcile.Result{}, nil
+}
+
+func (r *StorageClusterReconciler) logGrpcErrorAndReportEvent(instance *ocsv1.StorageCluster, grpcCallName string, err error, errCode codes.Code) {
+
+	cases := map[string]map[codes.Code]map[string]string{
+		"OnboardConsumer": {
+			codes.InvalidArgument: {
+				"msg":         "Token is invalid. Verify the token again or contact the provider admin",
+				"eventReason": "TokenInvalid",
+				"eventType":   corev1.EventTypeWarning,
+			},
+			codes.AlreadyExists: {
+				"msg":       "Token is already used. Contact provider admin for a new token",
+				"reason":    "TokenAlreadyUsed",
+				"eventType": corev1.EventTypeWarning,
+			},
+		},
+		"OffboardConsumer": {
+			codes.InvalidArgument: {
+				"msg":         "StorageConsumer UID is not valid. Contact the provider admin",
+				"eventReason": "UIDInvalid",
+				"eventType":   corev1.EventTypeWarning,
+			},
+		},
+		"UpdateCapacity": {
+			codes.InvalidArgument: {
+				"msg":         "StorageConsumer UID or requested capacity is not valid. Contact the provider admin",
+				"eventReason": "UIDorCapacityInvalid",
+				"eventType":   corev1.EventTypeWarning,
+			},
+			codes.NotFound: {
+				"msg":         "StorageConsumer UID not found. Contact the provider admin",
+				"eventReason": "UIDNotFound",
+				"eventType":   corev1.EventTypeWarning,
+			},
+		},
+		"GetStorageConfig": {
+			codes.InvalidArgument: {
+				"msg":         "StorageConsumer UID is not valid. Contact the provider admin",
+				"eventReason": "UIDInvalid",
+				"eventType":   corev1.EventTypeWarning,
+			},
+			codes.NotFound: {
+				"msg":         "StorageConsumer UID not found. Contact the provider admin",
+				"eventReason": "UIDNotFound",
+				"eventType":   corev1.EventTypeWarning,
+			},
+			codes.Unavailable: {
+				"msg":         "StorageConsumer is not ready yet. Will requeue after 5 second",
+				"eventReason": "NotReady",
+				"eventType":   corev1.EventTypeNormal,
+			},
+		},
+	}
+
+	eventDetails, ok := cases[grpcCallName][errCode]
+	if ok {
+		r.Log.Error(err, "External-OCS:"+grpcCallName+":"+eventDetails["msg"])
+		r.recorder.ReportIfNotPresent(instance,
+			eventDetails["eventType"], eventDetails["eventReason"], eventDetails["msg"])
+	}
```

Comment:
If this is not a global const map that the call sites utilize to figure out what to log then it would be preferable to just have a nested switch statement instead (or if statements). 
The current approach is just putting strain on the garbage collector because this very big map is created for each call and immediately after is thrown aside. It is also less readable (just look at the type definition).

I am valuing the simplest solution as it is always more readable and more maintainable, and in this case it is even less amount of code.


```suggestion
	msg := ""
	eventReason := ""
        eventType := ""

	if grpcCallName == OnboardConsumer {
		if errCode == codes.InvalidArgument {
			msg = "Token is invalid. Verify the token again or contact the provider admin";
			eventReason =  "TokenInvalid";
			eventType =  corev1.EventTypeWarning;
		} else if errCode == AlreadyExists {
			msg = "Token is already used. Contact provider admin for a new token";
			reason = "TokenAlreadyUsed";
			eventType = corev1.EventTypeWarning;
		}
		...
	} else if grpcCallName == OffboardConsumer {
	        ....
	} else if ....

        if msg != "" {
            r.Log.Error(err, "External-OCS:"+grpcCallName+":"+msg])
            r.recorder.ReportIfNotPresent(instance, eventType, eventReason,msg)
        }
```

Reply:
done